BitFaster.Caching/Lru/CustomPolicy.cs

```diff
@@ -0,0 +1,319 @@
+﻿using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace BitFaster.Caching.Lru
+{
+    /// <summary>
+    /// Defines a mechanism to calculate when cache entries expire.
+    /// </summary>
+    public interface IExpiryCalculator<K, V>
+    {
+        /// <summary>
+        /// Specify the inital time to live after an entry is created.
+        /// </summary>
+        TimeSpan GetExpireAfterCreate(K key, V value);
+
+        /// <summary>
+        /// Specify the time to live after an entry is read. The current TTL may be
+        /// be returned to not modify the expiration time.
+        /// </summary>
+        TimeSpan GetExpireAfterRead(K key, V value, TimeSpan currentTtl);
+
+        /// <summary>
+        /// Specify the time to live after an entry is updated.The current TTL may be
+        /// be returned to not modify the expiration time.
+        /// </summary>
+        TimeSpan GetExpireAfterUpdate(K key, V value, TimeSpan currentTtl);
+    }
+
+    /// <summary>
+    /// Defines a mechanism to determine the time to live for a cache item using function delegates.
+    /// </summary>
+    public readonly struct ExpiryCalculator<K, V> : IExpiryCalculator<K, V>

```

Comment:
Need to figure out how to represent this cleanly so that:

1. You can give a single delegate (ignoring current TTL)
2. You can give all 3 delegates

Easiest way is to make this a base class, but perf will be bad. Fastest way is 2 structs, but what should they be called?

`SimpleExpiryCalculator` and `ExpiryCalculator`?

Reply:
Removed delegate based impl from the API surface.